sql/analyzer/generate_index_scans.go

```diff
@@ -0,0 +1,267 @@
+		case *plan.IndexedTableAccess:
+			return false
+		case *plan.RecursiveCte:
+			// TODO: fix memory IndexLookup bugs that are not reproduceable in Dolt
+			// this probably fails for *plan.Union also, we just don't have tests for it
+			return false
+		case *plan.JoinNode:
+			// avoid changing anti and semi join condition indexes
+			return !n.Op.IsPartial() && !n.Op.IsFullOuter()
+		}
+
+		switch n := c.Parent.(type) {
+		// For IndexedJoins, if we are already using indexed access during query execution for the secondary table,
+		// replacing the secondary table with an indexed lookup will have no effect on the result of the join, but
+		// *will* inappropriately remove the filter from the predicate.
+		// TODO: the analyzer should combine these indexed lookups better
+		case *plan.JoinNode:
+			if n.Op.IsMerge() {
+				return false
+			} else if n.Op.IsLookup() || n.Op.IsLeftOuter() {
+				return c.ChildNum == 0
+			}
+		case *plan.TableAlias:
+			// For a TableAlias, we apply this pushdown to the
+			// TableAlias, but not to the resolved table directly
+			// beneath it.
+			return false
+		case *plan.Window:
+			// Windows operate across the rows they see and cannot
+			// have filters pushed below them. If there is an index
+			// pushdown, it will get picked up in the isolated pass
+			// run by the filters pushdown transform.
+			return false
+		case *plan.Filter:
+			// Can't push Filter Nodes below Limit Nodes
+			if _, ok := c.Node.(*plan.Limit); ok {
+				return false
+			}
+			if p, ok := c.Node.(*plan.Project); ok {
+				if _, ok := p.Child.(*plan.Limit); ok {
+					return false
+				}
+			}
+		}
+		return true
+	}
+
+	var handled []sql.Expression
+	return transform.NodeWithCtx(n, childSelector, func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
+		var ret sql.Node
+		var same transform.TreeIdentity
+		var err error
+		var lookup *indexLookup
+		var nameable sql.NameableNode
+		switch n := c.Node.(type) {
+		case *plan.Filter:
+			filtersByTable := getFiltersByTable(n)
+			filters := newFilterSet(n.Expression, filtersByTable, tableAliases)
+			filters.markFiltersHandled(handled...)
+			newF := removePushedDownPredicates(ctx, a, n, filters)
+			if newF == nil {
+				return FixFieldIndexesForExpressions(a, n, scope)
+			}
+			ret, _, err := FixFieldIndexesForExpressions(a, newF, scope)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+			return ret, transform.NewTree, nil
+		case *plan.TableAlias, *plan.ResolvedTable:
+			nameable = n.(sql.NameableNode)
+			ret, same, err, lookup = pushdownIndexesToTable(scope, a, nameable, indexes)
+		default:
+			return pushdownFixIndices(a, n, scope)
+		}
+		if err != nil {
+			return nil, transform.SameTree, err
+		}
+		if same {
+			return c.Node, transform.SameTree, nil
+		}
+		// TODO  mark lookup fields as used
+		handledF, err := getPredicateExprsHandledByLookup(ctx, a, nameable.Name(), lookup, tableAliases)
```

Comment:
It doesn't look like it, but this block is actually part of the `*plan.TableAlias, *plan.ResolvedTable` case (all the other cases return before here. Put it there instead

Reply:
Also do your error handling inside each case, not on the outside as here. Ditto for var declarations. Pull out a method if you need to